### PR TITLE
Pcie 151 pcie 152 pcie 153 pcie 154 pcie 155 pcie 156 pcie 157 pcie 158 endpoint dev implement controller class with all needed functions to handle any configuration request

### PR DESCRIPTION
**Pull request by mEmadEl-Dien**

### CODE REVIEW

**Comment by mEmadEl-Dien** [comment]
Branch name says a lot.

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+/**
+ * @brief Construct a new Configuration Controller and initializing some needed pointers to be used inside it 
+ */
+ConfigurationController::ConfigurationController()
+{
+    configuration = ConfigurationSpace::constructConfigurationSpace();
+    capability =  PCIECapability::constructPCIECapability();
+
+    handler = make_shared<ConfigurationRequestHandler>(configuration, capability);
+    tlpConstructor = make_shared<TLPConstructor>();
```
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:12]: why here you are constructing TLP, it is already implemented and you just use it?

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    cplUR = make_shared<CompletionWithUR>();
+}
+
+/**
+ * @brief Getting the Register number which we need to apply the Configuration request on
+ * 
+ * @param tlp -> The TLP coming to be handled
+ * @return int -> The Register Number to apply the Configuration request on
+ */
+int ConfigurationController::getRegisterNumber(TLP * tlp)
```
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:25]: this one need to be implemented before accepting this PR

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    return 0;
+}
+
+/**
+ * @brief Knowing whether the Configuration request carries a valid Register number to be handled or not
+ * 
+ * @param registerNumber -> Taken from the coming TLP
+ * @return int -> 1 for valid Register number, 0 otherwise
+ */
+int ConfigurationController::isValidRegisterNumber(int registerNumber)
```
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:39]: Okay so, this is blocked by another thing, try to finish the other one first and then we can merge this

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+ */
+unsigned int ConfigurationController::getTLPData(TLP * tlp)
+{
+    boost::dynamic_bitset<> dataPayloadBits;
+
+    unsigned int dataPayloadUint;
+
+    dataPayloadBits = tlp -> dataPayload;
+
+    dataPayloadUint = convertToUnsignedInt(dataPayloadBits);
```
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:73]: why you are converting it to unsigned Int ? 
I think you can store it in the register as Bitset normally

File: PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+ * 
+ * b) Configuration Write:
+ * Get Data to be written to the given register from the given TLP -> Perform the write operation inside the regiser, and if it's not a valid
+ * write operation construct a completion TLP with Unsupported request and return it, otherwise, Determine the TLP construction
+ * algorithm to be made (in this case will be Cpl) -> Perform this algorithm and return the TLP result 
+ * 
+ * @param tlp -> TLP to be handled
+ * @return TLP -> the resultant TLP
+ */
+TLP ConfigurationController::handleConfigurationRequest(TLP * tlp)
```
Comment (david-wagih) [inline PCIe 6.0 Packet Generator and Exerciser/PCIe 6.0 Packet Generator and Exerciser/Endpoint/controller_class/configuration_controller.cpp:122]: I have a question here 
why you created a handler for configuration requests, and still the logic of handling in the Configuration controller itself

**Review by david-wagih — COMMENTED** [review]
Very Nice work Mohamed 👍 
please check the Comments I attached.

**Pull request closed without merging**